src/lib/ares_parse_srv_reply.c

```diff
@@ -38,128 +38,80 @@ int
+    if (!srv_curr)
     {
-      ares_free (hostname);
-      return ARES_EBADRESP;
+      status = ARES_ENOMEM;
+      break;
     }
-  aptr += len + QFIXEDSZ;
-
-  /* Examine each answer resource record (RR) in turn. */
-  for (i = 0; i < ancount; i++)
+    if (srv_last)
     {
-      /* Decode the RR up to the data field. */
-      status = ares_expand_name (aptr, abuf, alen, &rr_name, &len);
-      if (status != ARES_SUCCESS)
-        {
-          break;
-        }
-      aptr += len;
-      if (aptr + RRFIXEDSZ > abuf + alen)
-        {
-          status = ARES_EBADRESP;
-          break;
-        }
-      rr_type = DNS_RR_TYPE (aptr);
-      rr_class = DNS_RR_CLASS (aptr);
-      rr_len = DNS_RR_LEN (aptr);
-      aptr += RRFIXEDSZ;
-      if (aptr + rr_len > abuf + alen)
-        {
-          status = ARES_EBADRESP;
-          break;
-        }
-
-      /* Check if we are really looking at a SRV record */
-      if (rr_class == C_IN && rr_type == T_SRV)
-        {
-          /* parse the SRV record itself */
-          if (rr_len < 6)
-            {
-              status = ARES_EBADRESP;
-              break;
-            }
-
-          /* Allocate storage for this SRV answer appending it to the list */
-          srv_curr = ares_malloc_data(ARES_DATATYPE_SRV_REPLY);
-          if (!srv_curr)
-            {
-              status = ARES_ENOMEM;
-              break;
-            }
-          if (srv_last)
-            {
-              srv_last->next = srv_curr;
-            }
-          else
-            {
-              srv_head = srv_curr;
-            }
-          srv_last = srv_curr;
-
-          vptr = aptr;
-          srv_curr->priority = DNS__16BIT(vptr);
-          vptr += sizeof(unsigned short);
-          srv_curr->weight = DNS__16BIT(vptr);
-          vptr += sizeof(unsigned short);
-          srv_curr->port = DNS__16BIT(vptr);
-          vptr += sizeof(unsigned short);
-
-          status = ares_expand_name (vptr, abuf, alen, &srv_curr->host, &len);
-          if (status != ARES_SUCCESS)
-            break;
-        }
-
-      /* Don't lose memory in the next iteration */
-      ares_free (rr_name);
-      rr_name = NULL;
-
-      /* Move on to the next record */
-      aptr += rr_len;
+      srv_last->next = srv_curr;
     }
+    else
+    {
+      srv_head = srv_curr;
+    }
+    srv_last = srv_curr;
 
-  if (hostname)
-    ares_free (hostname);
-  if (rr_name)
-    ares_free (rr_name);
-
-  /* clean up on error */
-  if (status != ARES_SUCCESS)
+    /* copy the host to newhost so we can free csrv_out */
+    char *newhost = NULL;
+    const char *tmphost = cares_srv_reply_get_host(csrv_curr);
+    if (!tmphost)
     {
-      if (srv_head)
-        ares_free_data (srv_head);
-      return status;
+      status = ARES_ENOMEM;
+      break;
     }
+    if ((newhost = malloc(strlen(tmphost) + 1)) != NULL) {
+      strcpy(newhost, tmphost);
+    }
+    srv_curr->host = newhost;
```

Comment:
Here is where the copy happens now.  Everything check out?

Reply:
1. Why not use strdup() ?
2. If the malloc fails, it must return error and fail the entire function call (without leaking anything)